cmd/notification.go

```diff
@@ -59,94 +59,126 @@ func (sys *NotificationSys) GetPeerRPCClient(addr xnet.Host) *PeerRPCClient {
 	return sys.peerRPCClientMap[addr]
 }
 
+// NotificationPeerErr returns error associated for a remote peer.
+type NotificationPeerErr struct {
+	Host xnet.Host // Remote host on which the rpc call was initiated
+	Err  error     // Error returned by the remote peer for an rpc call
+}
+
 // DeleteBucket - calls DeleteBucket RPC call on all peers.
-func (sys *NotificationSys) DeleteBucket(bucketName string) map[xnet.Host]error {
-	errors := make(map[xnet.Host]error)
+func (sys *NotificationSys) DeleteBucket(bucketName string) []NotificationPeerErr {
+	errs := make([]NotificationPeerErr, len(sys.peerRPCClientMap))
 	var wg sync.WaitGroup
+	idx := 1
 	for addr, client := range sys.peerRPCClientMap {
 		wg.Add(1)
-		go func(addr xnet.Host, client *PeerRPCClient) {
+		go func(idx int, addr xnet.Host, client *PeerRPCClient) {
 			defer wg.Done()
 			if err := client.DeleteBucket(bucketName); err != nil {
-				errors[addr] = err
+				errs[idx] = NotificationPeerErr{
+					Host: addr,
+					Err:  err,
+				}
 			}
-		}(addr, client)
+		}(idx, addr, client)
+		idx++
 	}
 	wg.Wait()
 
-	return errors
+	return errs
 }
 
 // SetBucketPolicy - calls SetBucketPolicy RPC call on all peers.
-func (sys *NotificationSys) SetBucketPolicy(bucketName string, bucketPolicy *policy.Policy) map[xnet.Host]error {
-	errors := make(map[xnet.Host]error)
+func (sys *NotificationSys) SetBucketPolicy(bucketName string, bucketPolicy *policy.Policy) []NotificationPeerErr {
+	errs := make([]NotificationPeerErr, len(sys.peerRPCClientMap))
 	var wg sync.WaitGroup
+	idx := 1
 	for addr, client := range sys.peerRPCClientMap {
 		wg.Add(1)
-		go func(addr xnet.Host, client *PeerRPCClient) {
+		go func(idx int, addr xnet.Host, client *PeerRPCClient) {
 			defer wg.Done()
 			if err := client.SetBucketPolicy(bucketName, bucketPolicy); err != nil {
-				errors[addr] = err
+				errs[idx] = NotificationPeerErr{
+					Host: addr,
+					Err:  err,
+				}
 			}
-		}(addr, client)
+		}(idx, addr, client)
+		idx++
 	}
 	wg.Wait()
 
-	return errors
+	return errs
 }
 
 // RemoveBucketPolicy - calls RemoveBucketPolicy RPC call on all peers.
-func (sys *NotificationSys) RemoveBucketPolicy(bucketName string) map[xnet.Host]error {
-	errors := make(map[xnet.Host]error)
+func (sys *NotificationSys) RemoveBucketPolicy(bucketName string) []NotificationPeerErr {
+	errs := make([]NotificationPeerErr, len(sys.peerRPCClientMap))
```

Comment:
What is the exact problem with existing code?  There is no way corrupt the map by those goroutines.  If you still paranoid, use `sync.Map`.

Handling slice is no difference by dealing with map here.

Reply:
I don't think so and I don't think you understand what maps is doing here - it cannot be used inside go-routine, look at the reference bug @balamurugana for the actual detail.  Slice is different because of the right positions and can be independently updated as a memory pointer. That is why i am not using `append` instead I have a slice with proper length equal to the peer clients. 

This is similar to all our code inside erasure code, where we update errs with their respective positions in go-routines.

Reply:
sync.Map is actually not necessary we don't need that kind of protection since these are all distinct calls to distinct peers @balamurugana - slices are correct here IMO which allow for no mutex requirement.  We can, in fact, do simply []error I brought the struct to satisfy error logging requirements added by the caller.

Reply:
> sync.Map is actually not necessary we don't need that kind of protection since these are all distinct calls to distinct peers @balamurugana

That's why I am talking about a map here.  Map will be nil or empty for all success and non-nil/entries only for errors.  Because data insertion happens only for failures.  Keeping N element slice is not required.

As overall parallel RPC calls to servers, currently we are not handling quorum based waiting for such goroutines all over the code.  This is wider problem to fix (As per my idea higher level type of N endpoints would solve cleanly).  I am seen many problems (without any logs) on failure tests.

As you are fixing map race, you could patch with `sync.Map`.  Let parallel RPC execution fix would take care automatically (Parallel RPC execution would return appropriate data structure for errors)

Reply:
> As overall parallel RPC calls to servers, currently we are not handling quorum based waiting for such goroutines all over the code. This is wider problem to fix (As per my idea higher level type of N endpoints would solve cleanly). I am seen many problems (without any logs) on failure tests.

Can you point to locations where we are not handling this? @balamurugana - From what I see I don't think this is true. 

> That's why I am talking about a map here. Map will be nil or empty for all success and non-nil/entries only for errors. Because data insertion happens only for failures. Keeping N element slice is not required.

If we are optimizing on memory then it is fine but our peer clients never go beyond 32 endpoints and it's premature I think. The problem I have with sync.Map is having to use mutex where it's not needed.

Reply:
Code places are wherever we do parallel execution of RPC calls.  As we wait for all RPC calls to finish,  the wait may be infinite.  It is wrong and client never gets response (As these bugs are already reported) or client will timeout itself.  The actual way of waiting should be done for quorum.  This will help to get response quicker for the client.

We don't need to worry about internal mutex or any locking mechanism of any Go builtin data structure (unless it has very serious performance penalty).

Reply:
> Code places are wherever we do parallel execution of RPC calls. As we wait for all RPC calls to finish, the wait may be infinite. It is wrong and the client never gets a response (As these bugs are already reported) or client will timeout itself. The actual way of waiting should be done for a quorum. This will help to get a response quicker for the client.

Where exactly do we do that? is it just erasure-createfile.go? can you point to other places in the codebase where we do that?

> We don't need to worry about internal mutex or any locking mechanism of any Go builtin data structure (unless it has very serious performance penalty).

Why not? how do we conclude sync.Map is correct here? if you look at Go's own source sync.Map is rarely used only in a handful of packages. 

> Point to this PR is any change it makes will be changed when parallel execution of RPC calls are handled properly. Less code change is preferred to avoid confusion in parallel RPC executions.

How is using sync.Map going to be minimal? how is that better understood than slices?